### PR TITLE
vreplication: table copying phase 1: create list of tables to copy

### PR DESCRIPTION
**Pull request by sougou**
We introduce two new states in vreplication: Init and Copying.
In the Init phase, one creates the vreplication row without a starting position. The VReplication will then copy tables from the source and keep them synced through vreplication. To achieve this, we have to identify the list of tables to copy. This PR is for this phase.

To find the list of tables to copy, we match the rules against the target (local) vttablet. Therefore, in order to be consistent, we don't send the same rules blindly to the source any more instead. Doing so may cause it to match a different set of tables than the target. Instead, we build an equivalent rule for each table that matched the source, and we send that to the source instead. This keeps the target and source consistent.

### CODE REVIEW

File: go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go [hunk 1]
```diff
@@ -74,26 +75,59 @@ const (
...
-			continue
-		}
-		sendRule, tablePlan, err := buildTablePlan(rule, tableKeys)
-		if err != nil {
-			return nil, err
+	for tableName := range tableKeys {
+		for _, rule := range filter.Rules {
+			switch {
+			case strings.HasPrefix(rule.Match, "/"):
+				expr := strings.Trim(rule.Match, "/")
```
Comment (rafael) [inline go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go]: The explanation for this change could be derived from the description in the PR. But this part of the code is tricky. I think starting to document in detail what's going on here, will be very helpful for our future selves.

File: go/vt/binlog/binlogplayer/binlog_player.go [hunk 2]
```diff
@@ -52,6 +52,10 @@ var (
 	// BlplTransaction is the key for the stats map.
 	BlplTransaction = "Transaction"
 
+	// VReplicationInit is for the Init state.
+	VReplicationInit = "Init"
```
[the remark this answers is not in the document]
Comment (dweitzman) [reply go/vt/binlog/binlogplayer/binlog_player.go]: There could be argument for defining the possible states in tabletmanagerdata.proto for reference by clients of the `VReplicationExec` API

File: go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go (hunk 1 above)
Comment (sougou) [reply go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go]: Added a high level comment for buildPlayerPlan.

File: go/vt/vttablet/tabletmanager/vreplication/vplayer.go [hunk 3]
```diff
@@ -91,55 +102,106 @@ func newVPlayer(id uint32, source *binlogdatapb.BinlogSource, sourceTablet *topo
...
+	defer vp.dbClient.Rollback()
+
+	// Check if table exists.
+	if _, err := vp.dbClient.ExecuteFetch("select * from _vt.copy_state limit 1", 10); err != nil {
+		// If it's a not found error, create it.
+		merr, isSQLErr := err.(*mysql.SQLError)
+		if !isSQLErr || !(merr.Num == mysql.ERNoSuchTable || merr.Num == mysql.ERBadDb) {
+			return err
+		}
+		log.Info("Looks like _vt.copy_state table may not exist. Trying to create... ")
```
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/vttablet/tabletmanager/vreplication/vplayer.go]: Removed.

File: go/vt/binlog/binlogplayer/binlog_player.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rafael) [reply go/vt/binlog/binlogplayer/binlog_player.go]: Cool. That makes sense. Let's create the protos after we deprecate.

**Pull request merged**